pytng/pytng.pyx

```diff
@@ -1037,314 +1137,301 @@ cdef class TNGFileIterator:
-        if stat != TNG_SUCCESS:
-            raise IOError("Critical failure: block cannot be read")
-        self._block_2d_numpy_cast(self.n_values_per_frame, self.n_atoms)
-        self.block_is_read = True
-
-    cdef void _refresh(self):
-        """Refreshes the TngDataBlock instance so that we can
-           read different blocks into the same instance"""
-        self.block_id = -1
-        self.step = -1
-        self.frame_time = -1
-        self.precision = -1
-        self.n_values_per_frame = -1
-        self.n_atoms = -1
-        self.block_is_read = False
-        # TODO this can be changed if the signature of MemoryWrapper is changed
-        self._wrapper = MemoryWrapper(1)
+    cpdef get_blockid(self, int64_t block_id, np.ndarray data):
+        """Get a block ID present at the current step and read it into a
+        NumPy array
 
-    @property
-    def values(self):
-        """the values of the block"""
-        if not self.block_is_read:
-            raise IOError(
-                'Values not available until block_read() has been called')
-        return self.values
-
-    cdef void _block_2d_numpy_cast(self, int64_t n_values_per_frame,
-                                   int64_t n_atoms):
-        """Casts the block to a 2D Numpy array whose lifetime is managed by
-           an associated MemoryWrapper instance"""
+        Parameters
+        ----------
+        block_id : int64_t
+           TNG block id to read from the current step
+        data : np.ndarray
+           NumPy array to read the data into, the required shape is determined
+           by the block dependency and the number of values per frame.
+        """
+        shape = data.shape
+        dtype = data.dtype
+
+        if dtype not in [np.int64, np.float32, np.float64]:
+            raise TypeError(
+                "PYTNG ERROR: datatype of numpy array not supported\n")
+
+        cdef void * values = NULL
+        cdef int64_t n_values_per_frame = -1
+        cdef int64_t n_atoms = -1
+        cdef double precision = -1
+        cdef char datatype = -1
+        cdef tng_function_status read_stat
 
-        if self.debug:
-            printf("PYTNG INFO: creating Numpy array \n")
-        if n_values_per_frame == -1 or n_atoms == -1:
-            raise ValueError(
-                "Array dimensions for block casting are not set correctly")
-        cdef int nd = 2
-        cdef int err
-        cdef npy_intp dims[2]
-        dims[0] = n_atoms
-        dims[1] = n_values_per_frame
-        self.values = PyArray_SimpleNewFromData(
-            2, dims, NPY_DOUBLE, self._wrapper.ptr)
-        Py_INCREF(self._wrapper)
-        err = PyArray_SetBaseObject(self.values, self._wrapper)
-        if err:
-            raise ValueError("Array object cannot be created")
-        if self.debug:
-            print(self.values)
+        cdef np.float32_t[:, :] _float_view
+        cdef int64_t[:, :] _int64_t_view
+        cdef double[:, :] _double_view
```

Comment:
Are any of these typed memoryviews going to hold C-contiguous data? You've probably thought of that already, but can maybe use the `::1` notation to squeeze out a little more perf under the hood.

Reply:
I think new structure resolves this, but would be grateful if you could take a quick peek. :)

Reply:
It certainly looks cleaner than the hard copy from before; I'll mark as resolved.